From: Ricky Benitez <rickyb@google.com>
Date: Mon, 5 Apr 2010 18:22:17 +0200
Subject: [PATCH] block: expose the statistics in blkio.time and blkio.sectors for the root cgroup
X-Git: a74b2adae06265b8cfa335d7d40d4a5abd11e977
References: FATE#311054
Patch-Mainline: 2.6.34

Currently, the io statistics for the root cgroup are maintained, but
they are not shown because the device information is not available at
the point that the root blkio cgroup is created. This patch updates
the device information when the statistics are updated so that the
statistics become visible.

Signed-off-by: Ricky Benitez <rickyb@google.com>
Acked-by: Vivek Goyal <vgoyal@redhat.com>
Signed-off-by: Jens Axboe <jens.axboe@oracle.com>
(cherry picked from commit d59fb281bd2ee811f9ba744ad073ac2991bae431)
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 block/cfq-iosched.c |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/block/cfq-iosched.c b/block/cfq-iosched.c
index 42028e7..08c7b58 100644
--- a/block/cfq-iosched.c
+++ b/block/cfq-iosched.c
@@ -944,6 +944,11 @@ cfq_find_alloc_cfqg(struct cfq_data *cfqd, struct cgroup *cgroup, int create)
 	unsigned int major, minor;
 
 	cfqg = cfqg_of_blkg(blkiocg_lookup_group(blkcg, key));
+	if (cfqg && !cfqg->blkg.dev && bdi->dev && dev_name(bdi->dev)) {
+		sscanf(dev_name(bdi->dev), "%u:%u", &major, &minor);
+		cfqg->blkg.dev = MKDEV(major, minor);
+		goto done;
+	}
 	if (cfqg || !create)
 		goto done;
 
-- 
1.6.0.2

